library/src/include/lib_device_helpers.hpp

```diff
@@ -94,19 +94,56 @@ enum copymat_direction
     copymat_from_buffer
 };
 
-template <typename T, typename U>
-ROCSOLVER_KERNEL void masked_copymat(copymat_direction direction,
+/** A mask that is always true. Typically used to make the mask optional,
+    by acting as the default when no other mask is provided. **/
+struct no_mask
+{
+    __device__ constexpr bool operator[](rocblas_int) const noexcept
+    {
+        return true;
+    }
+};
+
+/** An mask defined by an integer array (e.g., the info array) **/
+struct info_mask
+{
+    enum mask_transform
+    {
+        none,
+        negate
+    };
+
+    explicit constexpr info_mask(rocblas_int* mask, mask_transform transform = none) noexcept
+        : m_mask(mask)
+        , m_negate(transform == negate)
+    {
+    }
+
+    __device__ constexpr bool operator[](rocblas_int idx) const noexcept
+    {
+        return m_negate ^ !!m_mask[idx];
+    }
+
+    rocblas_int* m_mask;
+    bool m_negate;
+};
+
+/** COPY_MAT copies m-by-n array A into buffer if copymat_to_buffer, or buffer into A if copymat_from_buffer
+    An optional mask can be provided to limit the copy to selected matricies in the batch
+    If uplo = rocblas_fill_upper, only the upper triangular part is copied
+    If uplo = rocblas_fill_lower, only the lower triangular part is copied **/
+template <typename T, typename U, typename Mask = no_mask>
+ROCSOLVER_KERNEL void copy_mat(copymat_direction direction,
                                      const rocblas_int m,
                                      const rocblas_int n,
                                      U A,
                                      const rocblas_int shiftA,
                                      const rocblas_int lda,
                                      const rocblas_stride strideA,
                                      T* buffer,
-                                     const rocblas_int* mask,
+                                     const Mask mask = no_mask{},
```

Comment:
I don't see where you've defined a Mask type. How does that work?

Reply:
`Mask` is template type, like `T`. The concrete type would be `no_mask`, `info_mask`, or `rocblas_int*`. (I always wrapped things in `info_mask` in this PR to more closely match the existing code, but some of those probably could be left as `rocblas_int*`.)

Reply:
Ah, I see it now. Thanks!

Reply:
Just to clarify slightly, the fact that this is templated on the Mask type means that the `no_mask` and the `info_mask` versions of the template are separate kernels, directly corresponding to `copy_mat` and `masked_copymat`, respectively.

The template parameter is helpful for deciding how much you want to decide at compile-time vs. run-time. You can push things to compile-time by creating a custom Mask type specifically for that feature, or you can push the logic off until runtime by using a type that takes parameters (like the `info_mask` transform).